editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,590 @@
 #include "gradient_editor_plugin.h"
 
 #include "canvas_item_editor_plugin.h"
+#include "core/os/keyboard.h"
 #include "editor/editor_node.h"
 #include "editor/editor_scale.h"
 #include "editor/editor_settings.h"
+#include "editor/editor_spin_slider.h"
 #include "editor/editor_undo_redo_manager.h"
 #include "node_3d_editor_plugin.h"
+#include "scene/gui/color_picker.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/popup.h"
+#include "scene/gui/separator.h"
 
-void GradientReverseButton::_notification(int p_what) {
+int GradientEditorRect::get_point_at(int x) {
+	int result = -1;
+	int total_w = get_gradient_rect_width();
+	float min_distance = handle_width * 0.8; // Allow the cursor to be more than half a handle width away for ease of use.
+	for (int i = 0; i < gradient->get_point_count(); i++) {
+		// Ignore points outside of [0, 1].
+		if (gradient->get_offset(i) < 0) {
+			continue;
+		} else if (gradient->get_offset(i) > 1) {
+			break;
+		}
+		// Check if we clicked at point.
+		float distance = ABS(x - gradient->get_offset(i) * total_w);
+		if (distance < min_distance) {
+			result = i;
+			min_distance = distance;
+		}
+	}
+	return result;
+}
+
+int GradientEditorRect::predict_insertion_index(float p_offset) {
+	int result = 0;
+	while (result < gradient->get_point_count() && gradient->get_offset(result) < p_offset) {
+		result++;
+	}
+	return result;
+}
+
+int GradientEditorRect::get_gradient_rect_width() {
+	return get_size().width - get_size().height - draw_spacing - handle_width;
+}
+
+void GradientEditorRect::show_color_picker() {
+	ERR_FAIL_INDEX(selected_index, gradient->get_point_count());
+
+	picker->set_pick_color(gradient->get_color(selected_index));
+	Size2 minsize = popup->get_contents_minimum_size();
+	float viewport_height = get_viewport_rect().size.y;
+
+	// Determine in which direction to show the popup. By default popup below.
+	// But if the popup doesn't fit below and the Gradient Editor is in the bottom half of the viewport, show above.
+	bool show_above = get_global_position().y + get_size().y + minsize.y > viewport_height && get_global_position().y * 2 + get_size().y > viewport_height;
+
+	float v_offset = show_above ? -minsize.y : get_size().y;
+	popup->set_position(get_screen_position() + Vector2(0, v_offset));
+	popup->popup();
+}
+
+void GradientEditorRect::_color_changed(const Color &p_color) {
+	set_color(selected_index, p_color);
+}
+
+void GradientEditorRect::set_gradient(const Ref<Gradient> &p_gradient) {
+	gradient = p_gradient;
+	gradient->connect("changed", callable_mp((CanvasItem *)this, &CanvasItem::queue_redraw));
+}
+
+Ref<Gradient> GradientEditorRect::get_gradient() {
+	return gradient;
+}
```

Comment:
The GradientEditor class needed the gradient's metadata to update its buttons, but `set_gradient()` was too early. So I'm calling this at NOTIFICATION_READY.

Only thought of it now, but it might be better to instead store this metadata in two global variables?

Reply:
What do you mean by global variables?

Reply:
I thought of having a `snap_count` variable in GradientEditor too, which gets set in `set_gradient()` and then reused in `NOTIFICATION_READY`. On a second though, a variable that only gets used the before the control is ready doesn't sound good.